flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/security/token/HiveServer2DelegationTokenProvider.java

```diff
@@ -0,0 +1,235 @@
+                            try {
+                                LOG.info("Obtaining Kerberos security token for HiveServer2");
+
+                                String principal =
+                                        hiveConf.getTrimmed(
+                                                "hive.metastore.kerberos.principal", "");
+
+                                String tokenStr =
+                                        hive.getDelegationToken(
+                                                UserGroupInformation.getCurrentUser().getUserName(),
+                                                principal);
+                                Token<HiveServer2DelegationTokenIdentifier> hive2Token =
+                                        new Token<>();
+                                hive2Token.decodeFromUrlString(tokenStr);
+
+                                Credentials credentials = new Credentials();
+                                credentials.addToken(hive2Token.getKind(), hive2Token);
+
+                                HiveServer2DelegationTokenIdentifier tokenIdentifier =
+                                        getTokenIdentifier(hive2Token);
+
+                                if (tokenRenewalInterval == null) {
+                                    tokenRenewalInterval =
+                                            getTokenRenewalInterval(
+                                                    clock, tokenIdentifier, hive, tokenStr);
+                                }
+                                Optional<Long> validUntil =
+                                        tokenRenewalInterval.flatMap(
+                                                interval ->
+                                                        getTokenRenewalDate(
+                                                                clock, tokenIdentifier, interval));
+
+                                return new ObtainedDelegationTokens(
+                                        HadoopDelegationTokenConverter.serialize(credentials),
+                                        validUntil);
+
+                            } catch (Exception e) {
+                                LOG.error(
+                                        "Failed to obtain delegation token for {}",
+                                        this.serviceName(),
+                                        e);
+                                throw new FlinkRuntimeException(e);
+                            } finally {
+                                Hive.closeCurrent();
+                            }
+                        });
+    }
+
+    Optional<Long> getTokenRenewalInterval(
+            Clock clock,
+            HiveServer2DelegationTokenIdentifier tokenIdentifier,
+            Hive hive,
+            String tokenStr) {
+        Long interval;
+        LOG.debug("Got Delegation token is {} ", tokenIdentifier);
+        long newExpiration = getNewExpiration(hive, tokenStr);
```

Comment:
`getNewExpiration` is basically a one-liner.. why do we create a method for it?

Reply:
ok，I will change it